dsp/dsp.py

```diff
@@ -0,0 +1,254 @@
+    'boxcar',
+    'bartlett',
+    'hann',
+    'blackman',
+    'blackmanharris'
+]
+window_sizes = [64, 128, 256, 512, 1024, 2048, 4096]
+
+def compute_cola_window(config):
+    """
+    Using 'window_type' and 'window_size' values from 'config', computes 
+    desired window and corresponding hop size to achieve constant-overlap-add 
+    (COLA) property. Note that with some window - window size combinations the
+    quality of the COLA reconstruction is not perfect.
+    """
+    window_type = config['window_type']
+    window_size = config['window_size']
+    assert(window_size >= 64 and (window_size & (window_size - 1)) == 0)
+    if window_type == 'boxcar':
+        window = np.ones(window_size)
+        return window, window_size
+    window_size = window_size - 1
+    if window_type == 'bartlett':
+        window = sig.windows.bartlett(window_size)
+        overlap_factor = 0.5
+    elif window_type == 'hann':
+        window = sig.windows.hann(window_size)
+        overlap_factor = 0.5
+    elif window_type == 'blackman':
+        window = sig.windows.blackman(window_size)
+        overlap_factor = 2.0 / 3.0
+    elif window_type == 'blackmanharris':
+        window = sig.windows.blackmanharris(window_size)
+        overlap_factor = 0.75
+    else:
+        raise Exception(f"Unsupported window type {window_type}. Must be one of {', '.join(window_types)}")
+    
+    assert(overlap_factor >= 0.0 and overlap_factor <= 1.0)
+    
+    hop_factor = 1.0 - overlap_factor
+    initial_hop_size = int(np.round(hop_factor * window_size))
+
+    # COLA check
+    best_hop_size = initial_hop_size
+    best_gain_factor = 1.0
+    best_error = np.Infinity
+    for hop_size in [initial_hop_size, initial_hop_size - 1]:
+        num_windows = int(2.0 / hop_factor)
+        cola_check = np.zeros(num_windows * hop_size + (window_size - hop_size))
+        for i in range(num_windows):
+            n_start = i * hop_size
+            n_end = n_start + window_size
+            cola_check[n_start : n_end] += window
+        gain_factor = 1.0 / np.max(cola_check)
+        cola_check *= gain_factor
+
+        cola_start = window_size - hop_size // 2
+        cola_end = cola_start + window_size - hop_size
+        cola_check = cola_check[cola_start : cola_end]
+        cola_check /= np.max(cola_check)
+
+        error = np.sum(np.ones_like(cola_check) - cola_check)
+        if error < best_error:
+            best_error = error
+            best_hop_size = hop_size
+            best_gain_factor = gain_factor
+    
+    hop_size = best_hop_size
+    gain_factor = best_gain_factor
+    window = np.append(window * gain_factor, np.zeros(1))
+
+    return window, hop_size
+
+def get_windowed_frame_size(config):
+    window_size = config['window_size']
+    padding_factor = config['padding_factor']
+    frame_size = window_size * pow(2, padding_factor)
+    return frame_size
+
+def windowed_frame_analysis(signal, config):
+    window = config['window']
```

Comment:
I didn't see a `window` in the sample config file, but maybe I'm missing it. Which brings me to my main point which is that it seems that a lot of these asserts can be collected in a `validate` function which would validate things once after the config is loaded.

Reply:
Right now it's the user's responsibility to add `window` and `hop_size` to the config. This kind of goes against the pattern I was using overall, so as I said above it might make sense to change it so that `compute_cola_window` automatically adds it to the config. However I wanted to leave the option open of using arbitrary, 'non-COLA' `window` / `hop_size` settings. There isn't a `window` entry in the config file because the `compute_cola_window` method generates it

you can think of these various DSP operations as nodes in a directed graph. We need more / different information as we traverse it to specify the behavior of each node, and by letting these functions consume/pass along that information via the config object we're simplifying the process of composing that graph. 

again, I can see arguments against this approach and I do agree that it needs to be documented, but it does streamline things